File: ruby/native/HelloWorldKomencoNative/AutomatskiKomencoNative.rb
require 'json'
require 'net/http'
require 'uri'

require_relative 'QuantumCircuit'

include Automatski::Komenco

module Automatski
  module Komenco
    class AutomatskiKomencoNative
      def initialize(host, port)
        @host = host
        @port = port
      end

      def run(circuit, repetitions, topK)
        tstart = Time.now

        body = serialize_circuit(circuit, topK)
        struct = post_request(body)

        tend = Time.now
        execution_time = (tend - tstart) * 1000
        puts "Time Taken #{execution_time}ms"

        if struct.key?('error')
          puts struct['error']
          raise RuntimeError, struct['error']
        end

        deserialize_result(struct, repetitions)
      end

      private

      def serialize_circuit(circuit, topK)
        num_qubits = circuit.get_num_qubits

        operations = []
        measurements = []

        circuit.get_operations.each do |operation|
          gate = operation.get_gate
          params = operation.get_params
          qubits = operation.get_qubits

          if gate == 'measure'
            measurements.concat(qubits)
          else
            op = {
              'gate' => gate,
              'params' => params,
              'qubits' => qubits
            }
            operations << op
          end
        end

        puts "Executing Quantum Circuit With..."
        puts "#{num_qubits} Qubits And ..."
        puts "#{operations.size} Gates"

        if measurements.empty?
          raise RuntimeError, 'There are no measurements done at the end of the circuit.'
        end

        {
          'num_qubits' => num_qubits,
          'operations' => operations,
          'measurements' => measurements,
          'topK' => topK
        }
      end

      def post_request(body)
        url = URI.parse("http://#{@host}:#{@port}/api/komenco")
        http = Net::HTTP.new(url.host, url.port)
        request = Net::HTTP::Post.new(url.path, { 'Content-Type' => 'application/json', 'Accept' => 'application/json' })
        request.body = body.to_json

        response = http.request(request)
        if response.code.to_i >= 400
          raise StandardError, 'Error making POST request'
        end

        JSON.parse(response.body)
      end

      def deserialize_result(response_data, repetitions)
        measurements_strings = response_data['measurements']

        measurements = {}
        measurements_strings.each do |key, value|
          count = (value * repetitions).round
          measurements[key] = count
        end

        measurements
      end
    end
  end
end

File: c++/native/HelloWorldKomencoNative/AutomatskiKomencoNative.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <stdexcept>
#include <chrono>
#include <cmath>
#include <curl/curl.h>
#include <jsoncpp/json/json.h>
#include "QuantumCircuit.h"
#include "AutomatskiKomencoNative.h"

// Constructor
AutomatskiKomencoNative::AutomatskiKomencoNative(const std::string& host, int port)
    : host(host), port(port) {}

// Run method
std::unordered_map<std::string, int> AutomatskiKomencoNative::run(QuantumCircuit& circuit, int repetitions, int topK) {
    auto tstart = std::chrono::high_resolution_clock::now();

    Json::Value body = serializeCircuit(circuit, topK);
    Json::Value struct_ = postRequest(body);

    auto tend = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> executionTime = tend - tstart;
    std::cout << "Time Taken " << executionTime.count() << "ms\n";

    if (struct_.isMember("error")) {
        std::cout << struct_["error"].asString() << "\n";
        throw std::runtime_error(struct_["error"].asString());
    }

    return deserializeResult(struct_, repetitions);
}

// Serialize Circuit method
Json::Value AutomatskiKomencoNative::serializeCircuit(QuantumCircuit& circuit, int topK) {
    int numQubits = circuit.getNumQubits();

    Json::Value operations(Json::arrayValue);
    Json::Value measurements(Json::arrayValue);

    for (const auto& operation : circuit.getOperations()) {
        std::string gate = operation.getGate();
        std::vector<double> params = operation.getParams();
        std::vector<int> qubits = operation.getQubits();

        if (gate == "measure") {
            for (int qubit : qubits) {
                measurements.append(qubit);
            }
        } else {
            Json::Value op;
            op["gate"] = gate;
            for (double param : params) {
                op["params"].append(param);
            }
            for (int qubit : qubits) {
                op["qubits"].append(qubit);
            }
            operations.append(op);
        }
    }

    std::cout << "Executing Quantum Circuit With...\n";
    std::cout << numQubits << " Qubits And ...\n";
    std::cout << operations.size() << " Gates\n";

    if (measurements.empty()) {
        throw std::runtime_error("There are no measurements done at the end of the circuit.");
    }

    Json::Value result;
    result["num_qubits"] = numQubits;
    result["operations"] = operations;
    result["measurements"] = measurements;
    result["topK"] = topK;

    return result;
}

// Post Request method
Json::Value AutomatskiKomencoNative::postRequest(const Json::Value& body) {
    CURL* curl;
    CURLcode res;
    std::string readBuffer;

    curl = curl_easy_init();
    if (curl) {
        std::string url = "http://" + host + ":" + std::to_string(port) + "/api/komenco";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POST, 1L);

        struct curl_slist* headers = NULL;
        headers = curl_slist_append(headers, "Content-Type: application/json");
        headers = curl_slist_append(headers, "Accept: application/json");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        std::string jsonData = body.toStyledString();
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonData.c_str());

        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);

        res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            curl_easy_cleanup(curl);
            throw std::runtime_error("Error making POST request");
        }

        curl_easy_cleanup(curl);
    }

    Json::Value jsonResponse;
    Json::CharReaderBuilder readerBuilder;
    std::string errs;
    std::istringstream s(readBuffer);
    std::string doc;
    s >> doc;
    std::istringstream ss(doc);
    if (!Json::parseFromStream(readerBuilder, ss, &jsonResponse, &errs)) {
        throw std::runtime_error("Failed to parse JSON response");
    }

    return jsonResponse;
}

// Deserialize Result method
std::unordered_map<std::string, int> AutomatskiKomencoNative::deserializeResult(const Json::Value& responseData, int repetitions) {
    const Json::Value& measurementsStrings = responseData["measurements"];

    std::unordered_map<std::string, int> measurements;
    for (const auto& key : measurementsStrings.getMemberNames()) {
        int count = std::round(measurementsStrings[key].asDouble() * repetitions);
        measurements[key] = count;
    }

    return measurements;
}

// Write Callback method
size_t AutomatskiKomencoNative::WriteCallback(void* contents, size_t size, size_t nmemb, void* userp) {
    ((std::string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}